compressors/rANS.py

```diff
@@ -0,0 +1,335 @@
+    s, state = rans_base_decode_step(state)
+    ```
+- In the theoretical rANS version, the state keeps on increasing every time we call `rans_base_encode_step`
+  To make this practical, the rANS encoder ensures that after each encode step, the `state` lies in the acceptable range
+  `[L, H]`, where `L,H` are predefined interval values.
+
+  ```
+  state lies in [L, H], after every encode step
+  ```
+  To ensure this happens the encoder shrinks the `state` by streaming out its lower bits, *before* encoding each symbol. 
+  This logic is implemented in the function `shrink_state`. Thus, the full encoding step for one symbol is as follows:
+
+  ```python
+    ## Encoding one symbol
+    # output bits to the stream to bring the state in the range for the next encoding
+        state, out_bits = self.shrink_state(state, s)
+        encoded_bitarray = out_bits + encoded_bitarray
+
+    # core encoding step
+    state = self.rans_base_encode_step(s, state)
+  ```
+
+  The decoder does the reverse operation of `expand_state` where, after decoding a symbol, it reads in the a few bits to
+  re-map and expand the state to lie within the acceptable range [L, H]
+  Note that `shrink_state` and `expand_state` are inverses of each other
+
+  Thus, the  full decoding step 
+  ```python
+    # base rANS decoding step
+    s, state = self.rans_base_decode_step(state)
+
+    # remap the state into the acceptable range
+    state, num_bits_used_by_expand_state = self.expand_state(state, encoded_bitarray)
+  ```
+- For completeness: the acceptable range `[L, H]` are given by:
+  L = RANGE_FACTOR*total_freq
+  H = (2**NUM_BITS_OUT)*RANGE_FACTOR*total_freq - 1)
+  Why specifically these values? Look at the *Streaming-rANS* section on https://kedartatwawadi.github.io/post--ANS/
+
+
+## References
+1. Original Asymmetric Numeral Systems paper:  https://arxiv.org/abs/0902.0271
+2. Interactive write-up on rANS: https://kedartatwawadi.github.io/post--ANS/
+
+## TODO
+1. Implement alias-coding style rANS
+2. Adaptive rANS
+"""
+
+from dataclasses import dataclass
+import numpy as np
+from typing import Tuple, Any, List
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+from utils.bitarray_utils import BitArray, get_bit_width, uint_to_bitarray, bitarray_to_uint
+from core.data_block import DataBlock
+from core.prob_dist import Frequencies, get_mean_log_prob
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from utils.misc_utils import cache
+
+
+@dataclass
+class rANSParams:
+    """base parameters for the rANS encoder/decoder.
+    More details in the overview
+    """
+
+    # num bits used to represent the data_block size
+    DATA_BLOCK_SIZE_BITS: int = 32
+
+    # the encoder can output NUM_BITS_OUT at a time when it performs the state shrinking operation
+    NUM_BITS_OUT: int = 1  # number of bits
+
+    # rANS state is limited to the range [RANGE_FACTOR*total_freq, (2**NUM_BITS_OUT)*RANGE_FACTOR*total_freq - 1)]
```

Comment:
What is the tradeoff associated with RANGE_FACTOR? I can understand that it can't be too large because H is at most 32 bit (precision). On that note, is there a limit here on total_freq so we don't have overflow when doing `(2**NUM_BITS_OUT)*RANGE_FACTOR*total_freq` (python doesn't have int overflow but I'm talking more generally for a finite precision implementation)

Reply:
Does setting `NUM_BITS_OUT` to 8 lead to a range coding equivalent? Is that something people use?

Reply:
What is the importance of choosing `RANGE_FACTOR` (`t` from wiki) as something large rather than just 1. Is there precision loss on choosing it too small?

Reply:
in my understanding the main benefit of choosing `t` to be large is that the division during the encoding `x//freq[s]` is more precise. There is no other benefit I think

Reply:
Regarding `NUM_BITS_OUT` yes, people use 8bit, 16bit and 32bit "range" coder variants. that is precisely where the name `rANS` comes from. 

See: https://github.com/rygorous/ryg_rans

Reply:
<img width="1114" alt="image" src="https://user-images.githubusercontent.com/1708665/166401034-272ab309-0dda-47f8-a8e2-6ab0171f24c3.png">